service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -194,13 +198,22 @@ private[lagom] class TopicProducerActor[Message](
   private def eventsPublisherFlow(endpoints: String, offsetDao: OffsetDao) =
     Flow.fromGraph(GraphDSL.create(kafkaFlowPublisher(endpoints)) { implicit builder => publishFlow =>
       import GraphDSL.Implicits._
-      val unzip = builder.add(Unzip[Message, Offset])
+      val unzip = builder.add(Unzip[Option[Message], Offset])
+      val partition = builder.add(Partition[Option[Message]](2, {
+        case Some(_) => 0
+        case None    => 1
+      }))
+      val merge = builder.add(Merge[Any](2))
       val zip   = builder.add(Zip[Any, Offset])
       val offsetCommitter = builder.add(Flow.fromFunction { e: (Any, Offset) =>
         offsetDao.saveOffset(e._2)
       })
 
-      unzip.out0 ~> publishFlow ~> zip.in0
+      unzip.out0 ~> partition.in
+      partition.out(0).map(_.get) ~> publishFlow ~> merge.in(0)
+      partition.out(1) ~> merge.in(1)
+
+      merge.out ~> zip.in0
```

Comment:
I think instead of using a partition here, we could propagate the `Option[Message]` further and improve the `kafkaFlowPublisher` to branch (if `None` or `Some`) and use a [`ProducerMessage.PassThroughMessage`](https://doc.akka.io/docs/alpakka-kafka/current/atleastonce.html#publishing-to-message-dependent-topics) so nothing is emitted but the offset is committed.

Reply:
> I think instead of using a partition here, we could propagate the `Option[Message]` further and improve the `kafkaFlowPublisher` to branch (if `None` or `Some`) and use a [`ProducerMessage.PassThroughMessage`](https://doc.akka.io/docs/alpakka-kafka/current/atleastonce.html#publishing-to-message-dependent-topics) so nothing is emitted but the offset is committed.

I've been thinking about it and it may be a bit overkill to use `ProducerMessage.PassThroughMessage` (plus it would require at least one more allocation).